config/confighttp/compression.go

```diff
@@ -0,0 +1,110 @@
+package confighttp
+
+import (
+	"compress/gzip"
+	"compress/zlib"
+	"encoding/json"
+	"io"
+	"net/http"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+type errorHandler func(w http.ResponseWriter, r *http.Request, error error)
+
+type decompressor struct {
+	errorHandler errorHandler
+}
+
+type DecompressorOption func(d *decompressor)
+
+func WithErrorHandler(e errorHandler) DecompressorOption {
+	return func(d *decompressor) {
+		d.errorHandler = e
+	}
+}
+
+// HttpBodyDecompressor is a middleware that offloads the task of handling compressed
+// HTTP requests by identifying the compression format in the "Content-Encoding" header and re-writing
+// request body so that the handlers further in the chain can work on decompressed data.
+// It supports gzip and deflate/zlib compression.
+func HttpBodyDecompressor(h http.Handler, opts ...DecompressorOption) http.Handler {
+	d := &decompressor{
+		errorHandler: DefaultErrorHandler,
+	}
+	for _, o := range opts {
+		o(d)
+	}
+	return d.wrap(h)
+}
+
+func (d *decompressor) wrap(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		newBody, err := newBodyReader(r)
+		if err != nil {
+			d.errorHandler(w, r, err)
+			return
+		}
+		if newBody != nil {
+			defer newBody.Close()
+			// "Content-Encoding" header is removed to avoid decompressing twice
+			// in case the next handler(s) have implemented a similar mechanism.
+			r.Header.Del("Content-Encoding")
+			// "Content-Length" is set to -1 as the size of the decompressed body is unknown.
+			r.Header.Del("Content-Length")
+			r.ContentLength = -1
+			r.Body = newBody
+		}
+		h.ServeHTTP(w, r)
+	})
+}
+
+func newBodyReader(r *http.Request) (io.ReadCloser, error) {
+	switch r.Header.Get("Content-Encoding") {
+	case "gzip":
+		gr, err := gzip.NewReader(r.Body)
+		if err != nil {
+			return nil, err
+		}
+		return gr, nil
+	case "deflate", "zlib":
+		zr, err := zlib.NewReader(r.Body)
+		if err != nil {
+			return nil, err
+		}
+		return zr, nil
+	}
+	return nil, nil
+}
+
+// DefaultErrorHandler raises a 400 Bad request with error details in plain text.
+func DefaultErrorHandler(w http.ResponseWriter, _ *http.Request, err error) {
+	http.Error(w, err.Error(), http.StatusBadRequest)
+}
+
+// OTLPErrorHandler raises a 400 Bad request with a proto(google.rpc.Status) encoded error
+// as required by OTLP.
+func OTLPErrorHandler(w http.ResponseWriter, r *http.Request, err error) {
```

Comment:
(sorry reposting this for better visibility) 
I just noticed that the contrib library is using HTTPServerSettings.ToServer. I suppose it will be difficult to have functions that work for all use-cases, exposed via config. 

Eg: signalfxreceiver is using a [JSON encoded response.](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/master/receiver/signalfxreceiver/receiver.go#L198)

I wonder if it is better if we move this middleware to a common package eg: internal and have it re-used wherever it's required(I think to wrap grpc-gateway exposed muxes for now) inside the collector as opposed to exposing it via HTTPServerSettings?